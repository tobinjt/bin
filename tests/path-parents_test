#!/bin/bash

set -e -f -u -o pipefail

# shellcheck disable=SC1091
source ../path-parents

DIR="$(mktemp -d /tmp/path-parents-XXXXXXXXXX)"
STDOUT="${DIR}/stdout"
STDERR="${DIR}/stderr"
readonly DIR STDOUT STDERR

die() {
  echo "$@" >&2
  exit 1
}

# Run a command and check the exit code.
check_command_fails() {
  if main "$@" >> "${STDOUT}" 2>> "${STDERR}"; then
    die "Command succeeded: $*"
  fi
}
check_command_succeeds() {
  if ! main "$@" >> "${STDOUT}" 2>> "${STDERR}"; then
    die "Command failed: $*"
  fi
}

# Log a message to stdout and stderr logs to easily tell where output starts and
# ends.
start_test() {
  local file
  for file in "${STDOUT}" "${STDERR}"; do
    echo ">>> $* <<<" >> "${file}"
  done
}

test_main() {
  # Argument handling.
  start_test "No args"
  check_command_fails
  start_test "One filename"
  check_command_succeeds /etc/passwd
  start_test "Non-existent filename works"
  check_command_succeeds /qwerty/asdf/bar/foo/baz
  start_test "Only -h"
  check_command_succeeds -h
  start_test "Valid -s"
  check_command_succeeds -s 1 /etc/passwd
  start_test "Invalid -s"
  check_command_fails -s asdf /etc/passwd
  start_test "Unrecognised option"
  check_command_fails -q /etc/passwd

  # Path variants.
  start_test "Leading slashes"
  check_command_succeeds /etc/passwd
  check_command_succeeds -s 1 /etc/passwd
  start_test "No leading slashes"
  check_command_succeeds etc/passwd
  check_command_succeeds -s 1 etc/passwd
  start_test "Doubled slashes"
  check_command_succeeds etc//passwd
  check_command_succeeds -s 1 //etc/passwd
  check_command_succeeds -s 2 qwerty//asdf//bar/foo//baz

  start_test "Running out of components."
  check_command_succeeds -s 3 etc//passwd
  check_command_succeeds -s 10 /qwerty/asdf/bar/foo/baz

  start_test "Reading from stdin."
  echo "/qwerty/asdf/bar/foo/baz" | check_command_succeeds
  start_test "Reading from stdin and skipping."
  echo "/qwerty/asdf/bar/foo/baz" | check_command_succeeds -s 3

  diff -u path-parents_test.stdout "${STDOUT}"
  diff -u path-parents_test.stderr "${STDERR}"
  rm -r -f "${DIR}"
}

test_main
touch .path-parents_timestamp
