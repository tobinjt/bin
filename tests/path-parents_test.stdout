>>> No args <<<
>>> One filename <<<
/etc
/etc/passwd
>>> Non-existent filename works <<<
/qwerty
/qwerty/asdf
/qwerty/asdf/bar
/qwerty/asdf/bar/foo
/qwerty/asdf/bar/foo/baz
>>> Only -h <<<
Usage: /Users/johntobin/bin/path-parents [-s N] [PATH PATH . . .]
Usage: /Users/johntobin/bin/path-parents -h
If no filenames are provided reads from stdin.
Args:
  -h: Show usage message.
  -s N: Skip output of paths with N or fewer components (/bin has 1 component),
        /bin/test has 2 components.
>>> Valid -s <<<
/etc/passwd
>>> Invalid -s <<<
>>> Leading slashes <<<
/etc
/etc/passwd
/etc/passwd
>>> No leading slashes <<<
etc
etc/passwd
etc/passwd
>>> Doubled slashes <<<
etc
etc/passwd
/etc/passwd
qwerty/asdf/bar
qwerty/asdf/bar/foo
qwerty/asdf/bar/foo/baz
>>> Running out of components. <<<
>>> Reading from stdin. <<<
/qwerty
/qwerty/asdf
/qwerty/asdf/bar
/qwerty/asdf/bar/foo
/qwerty/asdf/bar/foo/baz
>>> Reading from stdin and skipping. <<<
/qwerty/asdf/bar/foo
/qwerty/asdf/bar/foo/baz
